### PR TITLE
Allow creating GCS remote state bucket with Bucket Policy Only

### PR DESCRIPTION
**Pull request by chrissng**
GCP recommends setting uniform bucket-level access (Bucket Policy Only) on GCS buckets, because it unifies and simplifies how one grants access to your Cloud Storage objects. If we regard terraform state to be sensitive, knowing that access control only is managed at the bucket-level helps to ensure that access to all objects is restricted to what is defined at the bucket-level.

This PR enables users to opt-in to this setting by defining `enable_bucket_policy_only = true` under `config`.

### CODE REVIEW

**Comment by brikis98** [comment]
@robmorgan Could you review this when you get a chance?

File: docs/_docs/02_features/keep-your-remote-state-configuration-dry.md [hunk 1]
```diff
@@ -202,10 +202,12 @@ remote_state {
 
  skip_bucket_versioning = true # use only if the object store does not support versioning
 
+ enable_bucket_policy_only = true # use only if uniform bucket-level access is needed (https://cloud.google.com/storage/docs/uniform-bucket-level-access)
```
Comment (robmorgan) [inline docs/_docs/02_features/keep-your-remote-state-configuration-dry.md]: should we default this to `false`?

File: docs/_docs/02_features/keep-your-remote-state-configuration-dry.md [hunk 2]
```diff
@@ -202,10 +202,12 @@ remote_state {
...
+
  encryption_key = "GOOGLE_ENCRYPTION_KEY"
 }
 ```
 
 If you experience an error for any of these configurations, confirm you are using Terraform v0.12.0 or greater.
 
-Further, the config options `gcs_bucket_labels` and `skip_bucket_versioning` are only valid for the backend `gcs`. They are used by terragrunt and are **not** passed on to terraform. See section [Create remote state and locking resources automatically](#create-remote-state-and-locking-resources-automatically).
+Further, the config options `enable_bucket_policy_only`, `gcs_bucket_labels` and `skip_bucket_versioning` are only valid for the backend `gcs`. They are used by terragrunt and are **not** passed on to terraform. See section [Create remote state and locking resources automatically](#create-remote-state-and-locking-resources-automatically).
```
Comment (robmorgan) [inline docs/_docs/02_features/keep-your-remote-state-configuration-dry.md]:
```suggestion
Further, the config options `gcs_bucket_labels`, `skip_bucket_versioning` and `enable_bucket_policy_only` are only valid for the backend `gcs`. They are used by terragrunt and are **not** passed on to terraform. See section [Create remote state and locking resources automatically](#create-remote-state-and-locking-resources-automatically).
```

Mirror the order above

File: docs/_docs/04_reference/config-blocks-and-attributes.md [hunk 3]
```diff
@@ -195,6 +195,7 @@ For the `s3` backend, the following additional properties are supported in the `
 
 For the `gcs` backend, the following additional properties are supported in the `config` attribute:
 
+- `enable_bucket_policy_only`: When `true`, the GCS bucket that is created to store the state will be configured to use uniform bucket-level access.
```
Comment (robmorgan) [inline docs/_docs/04_reference/config-blocks-and-attributes.md]: We should shift this under `skip_bucket_versioning` so its consistent with above.

File: docs/_docs/02_features/keep-your-remote-state-configuration-dry.md (hunk 1 above)
Comment (chrissng) [reply docs/_docs/02_features/keep-your-remote-state-configuration-dry.md]: Sure. fixed.

File: docs/_docs/04_reference/config-blocks-and-attributes.md (hunk 3 above)
Comment (chrissng) [reply docs/_docs/04_reference/config-blocks-and-attributes.md]: Got it

File: docs/_docs/02_features/keep-your-remote-state-configuration-dry.md (hunk 1 above)
Comment (robmorgan) [reply docs/_docs/02_features/keep-your-remote-state-configuration-dry.md]: Thanks 👍

**Comment by robmorgan** [comment]
I have a broken build, it looks like we need to update the Google Go SDK in order to set `BucketPolicyOnly`. I'll handle it.

**Comment by robmorgan** [comment]
Actually @chrissng would you be able to help me with this? We likely need to update the Go dependencies in the `Gopkg.toml` and `Gopkg.lock` files. You can test it successfully compiles using the following command: `go build -o terragrunt . && ./terragrunt`.

At the moment the [builds](https://circleci.com/gh/gruntwork-io/terragrunt/4016) are returning the following error:

```
# github.com/gruntwork-io/terragrunt/vendor/google.golang.org/api/internal
vendor/google.golang.org/api/internal/conn_pool.go:29:2: undefined: grpc.ClientConnInterface
FAIL	github.com/gruntwork-io/terragrunt/test [build failed]
?   	github.com/gruntwork-io/terragrunt/test/helpers	[no test files]
```

**Comment by chrissng** [comment]
@robmorgan I believe you may have attempted to update `google.golang.org/api` to the latest version. That vendor package seems to have an issue with grpc as you have pointed out. Version 0.16.0 of `google.golang.org/api` works.

I have also bumped `cloud.google.com/go` to `0.53.0`

**Comment by robmorgan** [comment]
I've kicked off another build in: https://circleci.com/gh/gruntwork-io/terragrunt/4109

**Comment by robmorgan** [comment]
It looks like the tests passed in https://circleci.com/workflow-run/9b72494c-de89-4065-a1b6-c09b78f626d8, so I'm going to merge and release this one.

**Pull request merged**